### PR TITLE
persist store paths, and migrate repo to v3

### PR DESCRIPTION
**Pull request by wdbaruni**
### Summary
This PR bumps the repo version from v2 to v3, where we introduce the following changes:
1. Change the execution store and job store default paths to no longer include the node name in the path. To clarify, v2 paths look like `~/.bacalhau/QmUBgU7xHKK44RuTHgrvnJfoSdZJS4fddT197iyTF5qjEV-compute/executions.db` and `QmUBgU7xHKK44RuTHgrvnJfoSdZJS4fddT197iyTF5qjEV-requester/jobs.db`, whereas in v3 they are `~/.baclahau/compute_store/executions.db` and `~/.bacalhau/orchestrator_store/jobs.db`. This change is needed to allow users to change our node names to longer depend on libp2p keys.
1. Persists the execution store and job store paths in `config.yaml` so that users won't lose their state if we change our default path names in the future
2. Adds a new `Node.Name` config, generates a node name still based on libp2p, and persist it. This will simplify an upcoming PR that allows users to define their own node name through the cli
3. Introduce migration scripts that will migrate v2 repos by creating `config.yaml` if it doesn't exist, persist the store paths, and rename old path names. No such migration will happen if the user provided their own `config.yaml` with store paths configured.

### `bacalhau config show`
I've also introduced `bacalhau config show` command to print out the current configuration which includes configs in `config.yaml`, env variables and default values.

### Few issues encountered
1. The repo is initialized in the `RootCmd` before any flags are registered. This means only default configurations are used when initializing the repo and any env variables or flags will be ignored. For example, if I run `bacalhau serve --compute-execution-store-path my-awesome-path` for the very first time, then bacalhau will still initialize the repo with default value, create `~/.bacalhau/compute-store` directory, and create `config.yaml` with `~/.bacalhau/compute-store` instead of my provided one. It is difficult to do things differently with our current setup as we initialize the repo in any command, but these flags are only available for `serve` command
2. We use global `viper` instance to set our config. I had to reset the config in my tests before each run. It might be safer to not use the global instance 
4. I've tried creating a repo per node in devstack, but couldn't get that to work because of our global configs. Configurations of the first repo initialized are also passed to the second repo's config, and we have test cases that fetch info of the very first repo initialized, such as https://github.com/bacalhau-project/bacalhau/blob/badd1ce5fd9508bd615f9a8e08bd8ce44d2eceb7/pkg/test/devstack/timeout_test.go#L112

## Summary by CodeRabbit

- **New Features**
	- Introduced new files to enhance configuration management, including displaying default and current settings.
	- Added functionality for handling private keys within the libp2p network.
	- Implemented a new approach for repository migration, including updating repository versions and handling specific configurations.
	- Enhanced CLI commands for better configuration setup and display.
- **Refactor**
	- Updated the logic flow in configuration setup, removing unnecessary imports and functions.
	- Simplified the initialization of repository and node configuration, including passing context parameters where needed.
	- Adjusted error messaging for clearer debugging.
	- Streamlined setup and teardown processes in tests.
- **Bug Fixes**
	- Fixed exclusion patterns in pre-commit hooks to correctly ignore specified directories.
- **Chores**
	- Updated imports and removed unused code across multiple files for cleaner codebase management.
- **Documentation**
	- Added summaries and comments for clarity in understanding the purpose and impact of changes.

### CODE REVIEW

File: pkg/repo/migrations/helpers.go
Comment (frrist) [inline pkg/repo/migrations/helpers.go:62]: Could move this to `pkg/lib` as `EqualSlices`

**Review by frrist — COMMENTED** [review]
Will approve once we are sure the changes around nodeID and nodeName don't effect compute <-> requester auth.

File: cmd/cli/serve/util.go
Comment (wdbaruni) [reply cmd/cli/serve/util.go:256]: Once a nodeID is generated, it will be persisted in `config.yaml` and will be maintained during restarts. If the user deletes the repo, or modifies the config file manually, then this will look like a new node joining the network, which is similar to if the user deletes the libp2p key today.

Is that still going to be a problem?

File: pkg/repo/migrations/helpers.go
Comment (wdbaruni) [reply pkg/repo/migrations/helpers.go:62]: I've copied this from the existing migration code. If I understand correctly, it checks if two _string_ slices are the same regardless of the order, as if order was also required, then `reflect.DeepEqual` should work. Not sure if the current implementation fits under `lib` as we will need to support generics or different types of slices for it to be useful as a lib. If you think it is still useful, then I can add it under `lib/collections` pkg

File: cmd/cli/serve/util.go
Comment (frrist) [reply cmd/cli/serve/util.go:256]: > Once a nodeID is generated, it will be persisted in config.yaml and will be maintained during restarts.

Where is this happening? I am missing where we update the config when the nodeID is set. 

Also, since node_name can be passed via the CLI as a flag or via an envvar wouldn't this mean if a user changed their node name with a flag I'd get a different auth key? I'd find that behavior unexpected.

File: pkg/repo/migrations/helpers.go
Comment (frrist) [reply pkg/repo/migrations/helpers.go:62]: Yeah or under an assertions package, its basically a copy of this: https://github.com/stretchr/testify/blob/v1.8.4/assert/assertions.go#L1039

File: cmd/cli/serve/util.go
Comment (wdbaruni) [reply cmd/cli/serve/util.go:256]: This is where it is happening https://github.com/bacalhau-project/bacalhau/blob/426-migrate-nodeID-boltdb-path/cmd/cli/serve/serve.go#L280

Yes, if the user run `bacalhau serve` with a different node name, then that is going to be a different node joining the network from the point of view of the requester. The mapping of executions to nodes is based on the node name/id, and will required different auth. Otherwise nodes can impersonate others by simply joining with their node names.

Though I am not sure how auth works today and how it is related to node names. Aren't we joining using a single nats token?

Comment (simonwo) [reply cmd/cli/serve/util.go:256]: > Once a nodeID is generated, it will be persisted in config.yaml and will be maintained during restarts.

That should meet the requirements of node-to-node auth, as long as we're happy that users changing the node name is equivalent to them creating a new node.

Tbh I think that is a little unexpected. What is the reason for allowing users to specifically set an ID? If it's to identify the nodes to humans, I would say that is conflating the node ID with its primary purpose, which is to identify nodes uniquely to the cluster. If we want to add human-readable identifiers, it would be more usual to do that in separate metadata.

Comment (simonwo) [reply cmd/cli/serve/util.go:256]: > Though I am not sure how auth works today and how it is related to node names. Aren't we joining using a single nats token?

The requester node will auto-generate a token that is expects to be used if one is not explicitly configured, and this token is generated by signing the node ID (a value unique to the node) using the node's private key (which only the node can do) – so that no other client can guess the token, the token is persistent, and there is little chance of a collision. As above, a user set node ID continues to allow this to work – changing the node ID (of the requester) would change the NATS auth token. If users want to migrate node IDs, they can just set an auth token explicitly (even the existing one).

Comment (frrist) [reply cmd/cli/serve/util.go:256]: Prior to this PR:
The node ID was derived from a private key: https://github.com/bacalhau-project/bacalhau/blob/main/cmd/cli/serve/serve.go#L386
Requester nodes generate compute API auth secrets (token) by signing their node ID with their private key: https://github.com/bacalhau-project/bacalhau/blob/main/cmd/cli/serve/util.go#L186.

The above means that compute API auth secrets will remain the same so long as the requester node uses the same private key. The private key of a node is used to derive its identity (ID). 

------------------------------

In this PR (and the one merging into it):
The NodeID (which we now also refer to as the NodeName) can be provided by the user via a config flag, environment variable, or name provider. This means the node ID is easily subject to change. https://github.com/bacalhau-project/bacalhau/blob/426-migrate-nodeID-boltdb-path/cmd/cli/serve/util.go#L257

The Requester nodes still generate the API auth secrets by signing their node ID with their private key: https://github.com/bacalhau-project/bacalhau/blob/426-migrate-nodeID-boltdb-path/cmd/cli/serve/util.go#L201

The result of this change means that even if the private key for a node is unchanged, the compute API auth token can change if a different flag is used, different config file is used, env var is set, provider is altered, VM IP changes, etc. i.e. the compute API auth secrete is no longer derived from a nodes identity. 

-----------------------

This change as coupled the identity of a node with its name and will result in confusing behavior around auth. I would like to see these things de-coupled before merging.

Comment (wdbaruni) [reply cmd/cli/serve/util.go:256]: Lets take few steps back. 


> Tbh I think that is a little unexpected. What is the reason for allowing users to specifically set an ID? If it's to identify the nodes to humans, I would say that is conflating the node ID with its primary purpose, which is to identify nodes uniquely to the cluster. If we want to add human-readable identifiers, it would be more usual to do that in separate metadata.

Allowing operators to define their own human readable names is a nice feature to help them debug and manage their clusters. I understand we can add as many metadata to the node, but the node name is what operators will see in the logs, metrics, who is running an execution and commands such as `bacalhau node get <name/id>`. This is the case with Kubernetes btw where they allow operators to override the node name from using the `hostname` to a user provided value using [`--hostname-override`](https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/#:~:text=%2D%2Dhostname%2Doverride%20string). 
Though I am happy to remove the cli flag to make it not easy to change the node name to avoid mistakes where operators will have to set it in the config file.


> The requester node will auto-generate a token that is expects to be used if one is not explicitly configured, and this token is generated by signing the node ID (a value unique to the node) using the node's private key (which only the node can do) – so that no other client can guess the token, the token is persistent, and there is little chance of a collision. As above, a persistent user set or random node ID continues to allow this to work – changing the node ID (of the requester) would change the NATS auth token. If users want to migrate node IDs but still allow existing nodes to connect using their previous token, they can just set an auth token explicitly (even the existing one).

I am missing the value of the requester node using its ID to generate the token. How is this different than the requester node generating a random token at startup, persisting it, and then reuse it after restart regardless if the requester node name changes or not? As I've mentioned in [Node ACL](https://github.com/bacalhau-project/expanso-planning/issues/463), we need to decouple how we identify our nodes (ie. names), from how they do auth to enable different auth methods. 

I understand we are creating secure networks by default where we generate a token if one is missing instead of making the network open by default, but that would only work for testing purposes as this won't work when you have multiple requesters. In that environment, users would have to manually define their auth by either generating a token outside of the requesters if we continue to only support tokens, or just one of the other auth methods provided by NATS. Actually auto creation of tokens is even complicating creating clusters for testing purposes as we are now back to the point where you have to create the requester node first, inspect its output, and then create the compute nodes, which is similar to when we relied on the libp2p peerID for bootstrapping.

Lastly, we should persist the AuthSecret after it was generated and avoid regenerating it on the fly each time. If we decide to change our implementation, such as use a different hash instead of sha256, then all existing networks will break. This is a similar problem I faced with this PR where changing the default path of jobstore will lost state of all existing networks, and I had to implement a migration that persists the existing paths in the config file.

Btw, I do believe changing the node name, should change its identify and auth, but that should be handled in NATS where the node name would map to a NATS user, and each user would have limited permission to the subjects it can subscribe and publish to.

> This change has coupled the identity of a node with its name and will result in confusing behavior around auth. I would like to see these things de-coupled before merging.

I agree with the problem, but not with the solution. Answered above that we shouldn't generate the token based on the nodeID. In fact I don't think we should be auto generating tokens at all :) 

Btw, this was discussed in last week's [engineering standup](https://www.notion.so/expanso/Engineering-Standup-ed22dd09186f4d6bbe5055d5c8f075c8?pvs=4#75c47e59335f4fe8b5f8f5a0bbeeaf8b), but I understand these issues may not be obvious from the beginning.

**Comment by wdbaruni** [comment]
btw, this discussion should be under https://github.com/bacalhau-project/bacalhau/pull/3450, and not this PR

**Comment by simonwo** [comment]
> Allowing operators to define their own human readable names is a nice feature to help them debug and manage their clusters. I understand we can add as many metadata to the node, but the node name is what operators will see in the logs, metrics, who is running an execution and commands such as bacalhau node get <name/id>. 

This is true – but it does not require us to conflate the system identifier with the human identifier. The more usual way is have an imutable system identifier and then allow humans to pick an optional human identifier. That way, we get [the best of both worlds](https://cloud.google.com/blog/products/api-management/api-design-choosing-between-names-and-identifiers-in-urls).

This is _actually_ the way that Docker works – each object (e.g. container, image) has a persistent and read-only system identifier, and also can be given an optional human identifier. The human identifier is preferred for printing in logs or tables, and the CLI commands always accept both. 

```
$ docker container ls
CONTAINER ID   IMAGE     COMMAND      CREATED          STATUS          PORTS     NAMES
2194f19a3330   ubuntu    "sleep 50"   12 seconds ago   Up 11 seconds             elated_chebyshev
$ docker container inspect elated_chebyshev
...
$ docker container inspect 2194f19a3330
...
```

Is there a good reason to not support that here?

> I am missing the value of the requester node using its ID to generate the token. How is this different than the requester node generating a random token at startup, persisting it, and then reuse it after restart regardless if the requester node name changes or not? 

It's simpler as it doesn't require the introduction of any further persistent state – but that's it. 

> Lastly, we should persist the AuthSecret after it was generated and avoid regenerating it on the fly each time. If we decide to change our implementation, such as use a different hash instead of sha256, then all existing networks will break.

Maybe, maybe not. If we're choosing to update to a different hashing algorithm, that is presumably for a good reason (i.e. there is now a trivial enough attack against the existing algorithm). In that case, whilst we could update the code that auto-generates tokens, we would still have a lot of users that had persistent auth tokens generated using a now insecure algorithim that is now attackable. So we'd have a choice between allowing those users to continue running their insecure servers without knowledge they are vulnerable, or to break their network and force them to upgrade to a less secure variant. It's not clear to me that the latter is worse than the former.

> Answered above that we shouldn't generate the token based on the nodeID. In fact I don't think we should be auto generating tokens at all :)

I've addressed in the other issue thread why we are auto-generating tokens. We are planning on the same thing for the client APIs, for the same reasons.

> Btw, this was discussed in last week's [engineering standup](https://www.notion.so/expanso/Engineering-Standup-ed22dd09186f4d6bbe5055d5c8f075c8?pvs=4#75c47e59335f4fe8b5f8f5a0bbeeaf8b), but I understand these issues may not be obvious from the beginning.

As I said in the standup, our existing code will continue to work fine as long as the node ID remains persistent and that we are happy that changing node ID will require clients to reconnect.

**Comment by wdbaruni** [comment]
> This is true – but it does not require us to conflate the system identifier with the human identifier. The more usual way is have an imutable system identifier and then allow humans to pick an optional human identifier. That way, we get [the best of both worlds](https://cloud.google.com/blog/products/api-management/api-design-choosing-between-names-and-identifiers-in-urls).

This is exactly what I am doing. In the [other PR](https://github.com/bacalhau-project/bacalhau/pull/3450),  I am generating a UUID as the node name by default and persist it, while enabling the user to chose their own if they want. Again this PR is not changing how we generate nodeID. I am still using libp2p key here, but I am just persisting the nodeID after it is generated.

> So we'd have a choice between allowing those users to continue running their insecure servers without knowledge they are vulnerable, or to break their network and force them to upgrade to a less secure variant. It's not clear to me that the latter is worse than the former.

I strongly advice that we follow a path where we can upgrade and improve things easily without breaking all users. It has been a pain to change how we name job store path in this PR mainly because we are computing the path always on the fly. If for some reason a network should not run using old auth methods, then we can always introduce a way to check and fail the node with proper error message instead of silently generating new keys.

> It's simpler as it doesn't require the introduction of any further persistent state – but that's it.

I understand. As mentioned above. I advice we still persist the token. This PR introduces a new way to easily persist certain configs ([here](https://github.com/bacalhau-project/bacalhau/blob/426-migrate-nodeID-boltdb-path/pkg/config/setters.go#L47)).

**Comment by wdbaruni** [comment]
Again, we are mixing a discussion that is not related to this PR and be under https://github.com/bacalhau-project/bacalhau/pull/3450. Are there any blockers for approving this PR?

File: cmd/cli/serve/util.go
Comment (frrist) [reply cmd/cli/serve/util.go:46]: >  I think it is a reasonable name as we are not modifying the user's disk or doing any complex initialization logic

`GetComputeConfig` calls `getExecutionStore` which is initializing a boltdb database and modifying the users disk.

**Comment by simonwo** [comment]
> I understand. As mentioned above. I advice we still persist the token. This PR introduces a new way to easily persist certain configs ([here](https://github.com/bacalhau-project/bacalhau/blob/426-migrate-nodeID-boltdb-path/pkg/config/setters.go#L47)).

I'd be happy for the token to be persisted – as I said, the only reason we didn't do that was because it was less state and complexity. I'd also be happy for it to be generated as e.g. a UUID in this case. But I'm still of the opinion that the ID should be separate from any kind of user-configurable name.

**Review by frrist — APPROVED** [review]
I also share Simons opinion that Node ID should be separate from user-configurable names. I would be very surprised if my node generated a new token simply because I deleted my config file - something that until recently has been optional.

In the interest of moving forward here I'll give this a :heavy_check_mark: and will assume NodeIDs will remain separate from user config.

**Pull request merged**
